### ISSUE TITLE
Add native darwin support

### ISSUE THREAD

**Issue by 05nelsonm** [05nelsonm/kmp-tor-binary#120]
Add `macOS`, `iOS`, `tvOS`, `watchOS` support.

Tor will need to be compiled, codesigned, and packaged so it can be extracted to the appropriate destination (just like Jvm/Js).

A new packaging tool and runtime library will be needed as there currently exists no way to add resources to a `.klib`. 

The simplest way it seems to package things is to base64 encode each resource and add as `.kt` files with some sort of header.

```kotlin
// resource_geoip.kt

// AUTO-GENERATED FILE
// DO NOT MODIFY

internal object resource_geoip: NativeResource(
    version = 1,
    filename = "geoip",
    rawSize = 1_039_091L,
    chunkSize = 4096,
    sha256 = "...",
    compression = NativeResource.Compression.GZIP,
    // ...
    chunks = Array<String>(254).apply {
        val a = this
        a[  0] = geoip_000; a[  1] = geoip_001
        // ...
        a[156] = geoip_156; a[157] = geoip_157
    },
)

// TODO: Maybe not a constant???
private const val geoip_000: String = """
......
4096 bytes of base64 encoded text
"""

private const val geoip_001: String = """
......
4096 bytes of base64 encoded text
"""
```

Will have to see if gzipping is possible before encoding. Also, this is a perfect use case for `EncoderDecoder.Feed` https://github.com/05nelsonm/encoding

**Comment by 05nelsonm** [comment]
I think that for the tool and it's use case, it'd work best as a gradle plugin. The whole process will go something like this

- Build `tor` via `./external/task.sh`
- Create detached signatures for darwin targets via `./external/task.sh`
- Package the `build` directory output (built binaries for given platform/arch). This will be 3 steps.
    - Apply detached code signatures
    - Move final binaries to directory `/library/binary/src/native_resources/<source set>` + `/io/matthewnelson/kmp/tor/binary`
        - e.g. `/library/binary/src/native_resources/iosArm64/io/matthewnelson/kmp/tor/binary/tor`
    - Gradle plugin will automatically check the `src/native_resources` directory for files and generate the `.kt` resource files.
    - So, the `tor` binary file will be picked up, and written in the aforementioned format to `src/iosArm64/io/matthewnelson/kmp/tor/binary/resource_tor.kt`

**Comment by 05nelsonm** [comment]
Split this issue up by creating #134 for implementing `NativeResource`

Adding the `blocked` label until that is completed.

**Comment by 05nelsonm** [comment]
Finally ready to start working on `task.sh` build script for darwin targets after #155

**Comment by 05nelsonm** [comment]
will need to use the `lipo` tool for darwin to make a multi-arch binary so they can be written to `iosMain`, `macosMain`, `tvosMain`, and `watchosMain` and not individually for each architecture. Will significantly reduce the size.

**Comment by 05nelsonm** [comment]
> will need to use the `lipo` tool for darwin to make a multi-arch binary so they can be written to `iosMain`, `macosMain`, `tvosMain`, and `watchosMain` and not individually for each architecture. Will significantly reduce the size.

Unsure how the simulators will work with this because there'll be multiple arm64 architectures. May need to configure source sets separately.

e.g.
```
darwinMain
    |--- iosMain
    |        |--- iosArm64
    |         '--- iosX64
    |--- iosSimulatorArm64
    ...
```

**Comment by 05nelsonm** [comment]
So compiling on `macOS` machine using `Xcode.app` is simply not going to work. Output is not reproducible. Working on another docker container for `iPhoneOS` and `iPhoneSimulator` SDKs.

**Comment by 05nelsonm** [comment]
`ios`/`ios-simulator` docker files will live in `kmp-tor-binary` until things are ready for release. #168 has been created so I do not forget about it :smirk_cat:

**Comment by 05nelsonm** [comment]
Native support for `macOS` was implemented in #171 

Remaining targets will be rolled out separately (tons of work getting things compiling via docker using `cctools-port`). Closing this as "completed" and delegating the remaining items to the following issue tickets:

- #175 
- #176
- #177
